pkg/controllers/chart/chart.go

```diff
@@ -831,3 +894,38 @@ func hashObjects(job *batch.Job, objs ...metav1.Object) {
 func setBackOffLimit(job *batch.Job, backOffLimit *int32) {
 	job.Spec.BackoffLimit = backOffLimit
 }
+
+func deepMergeStruct(target, source interface{}) {
+	targetVal := reflect.ValueOf(target).Elem()
+	sourceVal := reflect.ValueOf(source).Elem()
+
+	for i := 0; i < targetVal.NumField(); i++ {
+		targetField := targetVal.Field(i)
+		sourceField := sourceVal.Field(i)
+
+		if !sourceField.IsNil() {
+			if targetField.IsNil() {
+				targetField.Set(sourceField)
+			} else {
+				// If fields are pointers and Structs, merge recursively
+				if targetField.Kind() == reflect.Ptr && targetField.Elem().Kind() == reflect.Struct &&
+					sourceField.Kind() == reflect.Ptr && sourceField.Elem().Kind() == reflect.Struct {
+					deepMergeStruct(targetField.Interface(), sourceField.Interface())
+				} else {
+					// Otherwise, overwrite target field with source field value
+					targetField.Set(sourceField)
+				}
+			}
+		}
+	}
+}
+
+func setSecurityContext(job *batch.Job, chart *v1.HelmChart) {
+	if chart.Spec.PodSecurityContext != nil {
+		deepMergeStruct(job.Spec.Template.Spec.SecurityContext, chart.Spec.PodSecurityContext)
```

Comment:
I'm still not understanding why this is necessary. If chart.Spec.PodSecurityContext/SecurityContext is set, shouldn't we just set that value in the chart, and skip all this deep copying and merging? I've not seen this sort of behavior as necessary in any of our other controllers that allow setting these fields from a higher level spec.

Reply:
The deep merge is necessary because the `podSecurityContext` and `securityContext` are both maps. If the user would like to change just a single field, he/she would have to define all the default fields plus his/her changed field. With the deep merge, the defaults are _updated_ with the user-specified fields recursively. We could remove the deep merge, but I think it's more user friendly if we keep it. I guess, you didn't need any deep copy in other controllers because those controllers don't use maps but only primitive types (string, bool, ...)?

Reply:
> The deep merge is necessary because the podSecurityContext and securityContext are both maps.

No they're not? They're structs.

https://pkg.go.dev/k8s.io/api/core/v1#SecurityContext
https://pkg.go.dev/k8s.io/api/core/v1#ContainerSecurityContext

> If the user would like to change just a single field, he/she would have to define all the default fields plus his/her changed field.

Yes. The expectation is generally that the field from the spec **sets** the field in the job, not that it is merged with the defaults.  See for example:
https://github.com/rancher/system-upgrade-controller/blob/af373602d649f108321eb4c784d490972499ad9f/pkg/upgrade/container/container.go#L34-L38

Reply:
> No they're not? They're structs.

In the YAML terminology it's called a [mapping or `map`](https://yaml.org/spec/1.2.2/#mapping).

> Yes. The expectation is generally that the field from the spec **sets** the field in the job, not that it is merged with the defaults. If the user wants to extend the defaults, they must explicitly include them. See for example: https://github.com/rancher/system-upgrade-controller/blob/af373602d649f108321eb4c784d490972499ad9f/pkg/upgrade/container/container.go#L34-L38

I understand. The problem of having to set the `securityContext` always as fully defined map is that if the user wants to change just a single field and keep all the other fields as they are set by default, the user must copy&paste _all_ the fields instead of just setting the fields of interest. The deep merge principle aligns with how Helm charts work - modification to a single field in a map _updates_ the default map values set in the `values.yaml` file.

I can remove the deep merge if would like to maintain compatibility with your other controllers. Please just let me know.

Reply:
yes, but we're not dealing with yaml here - we're dealing with golang structs.

The expected behavior for our controllers is that fields set in the spec override defaults, instead of merging with them. That's what I would look for here.

Reply:
It should be fixed in the last commit.